### PR TITLE
CSI: Add secrets flag support for delete volume

### LINKED ISSUE #10997: CSI: Add secrets support for delete volume
### Proposal
Add the option to provide a secret when deleting a volume, i.e.

```
nomad volume delete vol1 --secrets a=b,c=d
```

Similar to https://github.com/hashicorp/nomad/issues/10640

These secrets will take precedence over any stored volume secrets. They'll be combined into a single secrets map and sent to the CSI driver.

I can work on this feature if the proposal is accepted.

### Use-cases
For cases when a new secret must be provided to the storage system during deletion. I.e. if a token has expired or encryption key change.

### Attempted Solutions
n/a

### PR DESCRIPTION
**Pull request by ggriffiths**
Fixes #10997 
Signed-off-by: Grant Griffiths <ggriffiths@purestorage.com>

### CODE REVIEW

**Comment by ggriffiths** [comment]
Tested this on my cluster. The scenario is that a PX volume has been created with a token that expires in one minute. 
```
[root@ip-70-0-139-170 nomadspecs]# nomad volume delete pxvol1 
Error deleting volume: Unexpected response code: 500 (controller delete volume: rpc error: controller delete volume: CSI.ControllerDeleteVolume: controller plugin returned an internal error, check the plugin allocation logs for more information: rpc error: code = Internal desc = Unable to delete volume with id 282299434038939129: rpc error: code = PermissionDenied desc = Token is expired)
```

Because the stored token has expired as seen above, we must pass generate a new token:
```
[root@ip-70-0-139-170 nomadspecs]# /opt/pwx/bin/pxctl auth token generate --auth-config admin.yaml --shared-secret mysecret --issuer openstorage.io --token-duration=1m
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImdyYW50QG9wZW5zdG9yYWdlLmlvIiwiZXhwIjoxNjMyOTY2OTgzLCJncm91cHMiOlsiKiJdLCJpYXQiOjE2MzI5NjY5MjMsImlzcyI6Im9wZW5zdG9yYWdlLmlvIiwibmFtZSI6IkdyYW50Iiwicm9sZXMiOlsic3lzdGVtLmFkbWluIl0sInN1YiI6ImdyYW50QG9wZW5zdG9yYWdlLmlvIn0.g4qvXrq8kfDL1BhmKr5YYvInXwwZROLbW58XRm_uEtM
```

After that, pass it to the volume delete command:
```
[root@ip-70-0-139-170 nomadspecs]# nomad volume delete --secrets=auth-token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImdyYW50QG9wZW5zdG9yYWdlLmlvIiwiZXhwIjoxNjMyOTY2OTgzLCJncm91cHMiOlsiKiJdLCJpYXQiOjE2MzI5NjY5MjMsImlzcyI6Im9wZW5zdG9yYWdlLmlvIiwibmFtZSI6IkdyYW50Iiwicm9sZXMiOlsic3lzdGVtLmFkbWluIl0sInN1YiI6ImdyYW50QG9wZW5zdG9yYWdlLmlvIn0.g4qvXrq8kfDL1BhmKr5YYvInXwwZROLbW58XRm_uEtM pxvol1
Successfully deleted volume "pxvol1"!
```

**Comment by ggriffiths** [comment]
A few flakey tests seem to be failing, but I think that's unrelated. Ready for review

**Comment by lgfa29** [comment]
Thank you @ggriffiths! I review this soon 🙂

File: command/volume_delete.go [hunk 1]
```diff
@@ -78,8 +83,8 @@ func (c *VolumeDeleteCommand) Run(args []string) int {
 
 	// Check that we get exactly two arguments
 	args = flags.Args()
-	if l := len(args); l != 1 {
-		c.Ui.Error("This command takes one argument: <vol id>")
+	if l := len(args); l < 1 {
+		c.Ui.Error("This command takes at least one argument: <vol id>")
```
Comment (lgfa29) [inline command/volume_delete.go:87]: Is this change necessary? I think `flags.Args()` returns only the arguments to the command, not flags. So the number of expected arguments is still 1 (`<vol id>`).

File: api/csi.go [hunk 2]
```diff
@@ -101,8 +101,13 @@ func (v *CSIVolumes) Create(vol *CSIVolume, w *WriteOptions) ([]*CSIVolume, *Wri
...
 // that's already been deregistered can be deleted.
-func (v *CSIVolumes) Delete(externalVolID string, w *WriteOptions) error {
-	_, err := v.client.delete(fmt.Sprintf("/v1/volume/csi/%v/delete", url.PathEscape(externalVolID)), nil, w)
+func (v *CSIVolumes) Delete(externalVolID string, secrets string, w *WriteOptions) error {
+	qp := url.Values{}
+	if secrets != "" {
+		qp.Set("secrets", secrets)
+	}
+
+	_, err := v.client.delete(fmt.Sprintf("/v1/volume/csi/%v/delete?%s", url.PathEscape(externalVolID), qp.Encode()), nil, w)
```
Comment (lgfa29) [inline api/csi.go:110]: Hum....passing secret values by HTTP query params is _not great_ as they can get accidentally stored in log files. I see that we do it for other CSI endpoints, so totally not your fault 🙂 

I will try to get a PR to accept secrets as a header instead.

File: command/volume_delete.go [hunk 3]
```diff
@@ -70,6 +73,8 @@ func (c *VolumeDeleteCommand) Name() string { return "volume delete" }
 func (c *VolumeDeleteCommand) Run(args []string) int {
 	flags := c.Meta.FlagSet(c.Name(), FlagSetClient)
 	flags.Usage = func() { c.Ui.Output(c.Help()) }
+	var secrets string
```
Comment (lgfa29) [inline command/volume_delete.go:76]: We usually store flags in the command `struct`: https://github.com/hashicorp/nomad/blob/v1.1.6/command/job_status.go#L23-L26

File: website/content/docs/commands/volume/delete.mdx [hunk 4]
```diff
@@ -16,7 +16,7 @@ deleted.
 ## Usage
 
 ```plaintext
-nomad volume delete [options] [volume]
+nomad volume delete [options] [volume] [-secrets key=value,key2=value2]
```
Comment (lgfa29) [inline website/content/docs/commands/volume/delete.mdx:19]:
```suggestion
nomad volume delete [options] [volume]
```

`-secrets` is a flag, so it's covered by the `[options]` placeholder.

**Comment by tgross** [comment]
@ggriffiths I've just merged https://github.com/hashicorp/nomad/pull/12144 which has that change to use HTTP headers. If you'd like to rebase this PR on `main` I'd be happy to help get it pushed thru review. Or if you'd like I can carry the PR from here and get it merged in.

**Comment by ggriffiths** [comment]
Cool, I'll rebase this PR. Thanks @tgross !

**Comment by tgross** [comment]
Hi @ggriffiths! We're getting very close to releasing the Nomad 1.3.0 beta, so I've carried this PR to rebase it and bring it up to date with the other secrets-related PRs we've done recently. I'm going to make sure this is green in CI and then probably tap @lgfa29 for a review of the extra commits I've added before we get it merged in.

**Comment by ggriffiths** [comment]
sounds good, thanks @tgross!

**Review by lgfa29 — APPROVED** [review]
Glad to see this moving forward, thank you for taking care of it @tgross and sorry for the delay in getting this merged @ggriffiths.

**Comment by ggriffiths** [comment]
Not a problem @lgfa29, thanks!
